### ISSUE TITLE
Update `ModuleFactory` to parse based on `File` instead of `filename`

### ISSUE THREAD

**Issue by jayasuryat** [jayasuryat/mendable#31]
[`ModuleFactory`](https://github.com/jayasuryat/mendable/blob/a7ee7c9b129b5849fcfa55f259794281a5b7ea49/scanner/src/main/java/com/jayasuryat/mendable/scanner/ModuleFactory.kt#L27) right now takes a `String` as an input to parse a `Module` object which is included as part of all the Mendable reports.

While this is sufficient to parse `Module` from just the file name, this might not be sufficient in the context of nested modules. Modules could be nested in other folders like:
<img width="169" alt="Screenshot 2023-08-22 at 10 56 02 PM" src="https://github.com/jayasuryat/mendable/assets/37530409/05279bc1-22a8-49e9-8448-0c72a4b2f23e">

In the case of the image above there are three `modules` nested in each of the two `root folders` (`features-v2` & `feature_libs`). As the names of the nested modules are same, Compose compiler will end up generating metrics files with the same name (as the current behavior does not consider folder nesting in the naming scheme).
> Or worse, if the target directory is the same for all, then it might end up replacing the generated metrics files.

So in such cases just relying on `filename` does not give enough control or information required to derive the desired `module name` appropriately.

---
### TLDR: 
 So, instead of using `filename` to parse a `Module` object, it would be better to parse it based on the `File` itself.

A `File` already has the `filename`, so all of the existing logics are covered, and it also has `path` which could be used further nail down module name more appropriately, especially in cases like the one described above (by comparing the file path with the root project's path for example).

Yes, this would also be a backwards-incompatible change

---

### TODO:
- [x] Update signature of `ModuleFactory`
- [x] Update implementation of `DefaultModuleFactory`
- [x] Update all of the usages (if any)
- [x] Update tests
- [x] Dump and validate API

**Comment by Akshayp02** [comment]
Hey i would like to work on this issue can you assign me this to explore more.

**Comment by jayasuryat** [comment]
Hey @Akshayp02 👋
Sorry for the late response, somehow I missed this notification. 

---

### This issue is still up for grabs, and if you are still interested you can take this up. Let me know.

**Comment by Akshayp02** [comment]
Hello sir, Thanks for the reply. 😊
 sir I'm still interested can you assign me this issue so i can explore it.💫

**Comment by jayasuryat** [comment]
Sure thing @Akshayp02, I've assigned this issue to you.
More-or-less of what has to be done is listed in the TODO section of the [original comment](https://github.com/jayasuryat/mendable/issues/31#issue-1861910787). You can continue from there.

> ### TODO:
> * [ ]  Update signature of `ModuleFactory`
> * [ ]  Update implementation of `DefaultModuleFactory`
> * [ ]  Update all of the usages (if any)
> * [ ]  Update tests
> * [ ]  Dump and validate API

**Comment by Akshayp02** [comment]
Sure sir

**Comment by jayasuryat** [comment]
No need to call me sir @Akshayp02, you can just call me @jayasuryat 😊

**Comment by jayasuryat** [comment]
Hi @Akshayp02,
I hope the steps explained [here](https://github.com/jayasuryat/mendable/discussions/32#discussioncomment-7059956) helped you with running the project.

Were you able to run the project and test your changes?

**Comment by jayasuryat** [comment]
Hi, @Akshayp02.

I'm planning to release [`v0.7.0`](https://github.com/jayasuryat/mendable/milestone/3) this week. And this feature would be a part of that release too. 

Please let me know if you were able to work on this issue. As the release is coming up shortly, If you are unable to take care of this right now, no worries, I'll assign this back to myself and take care of it. Let me know.

**Comment by Akshayp02** [comment]
Good morning @jayasuryat  💫
Thanks for your patience and Sorry for the delay actually due to my academic I was not active GitHub that's why it's delayed once again sorry for that I'll  complete it within 2-3 days. Actually my exams are going on so that's why I need 2-3 days to do it.

**Comment by jayasuryat** [comment]
Good morning, @Akshayp02.

No worries, and all the very best for your exams. 
Don't take pressure about this. It is fine and I understand.

I'll get back to you in a couple of days when your exams are done.

**Comment by Akshayp02** [comment]
Thank you so much ❤

**Comment by jayasuryat** [comment]
Hi, @Akshayp02!

I hope your exams went well, and I hope you found a moment to address this issue.

I'm currently quite pressed for time, @Akshayp02. I hope you can understand the situation. If possible, could you please complete this by tomorrow? Otherwise, regrettably, I may have to handle it myself, as this release is causing delays in other areas.

**Comment by Akshayp02** [comment]
Hello @jayasuryat  actually my exams are till Saturday. I'm sorry for the inconvenience but there are time limitations so please you can provide forword if there is any section where I can contribute so please let me know.

**Comment by jayasuryat** [comment]
No worries, I'll take care of it. Thanks for understating @Akshayp02.

Whenever you have time the next time, you can checkout issues with [`good-first-issue`](https://github.com/jayasuryat/mendable/issues?q=is%3Aopen+label%3Agood-first-issue+no%3Aassignee) label if you'd like to contribute.

Again, all the very best for your exams.
